From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alpha <alphakr93@outlook.com>
Date: Wed, 19 Jan 2022 23:18:02 +0900
Subject: [PATCH] Do not send useless entity packets

Original by William Blake Galbreath <Blake.Galbreath@GMail.com>
(PurpurMC Team)

Copyright (c) 2022 PurpurMC LLC

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index 3441339e1ba5efb0e25c16fa13cb65d2fbdafc42..b5fc0315a43d17857845e0d0e91d3f4a744d1c2c 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -181,6 +181,7 @@ public class ServerEntity {
                         this.teleportDelay = 0;
                         packet1 = new ClientboundTeleportEntityPacket(this.entity);
                     }
+                    if (net.prismarineteam.prismarine.PrismarineConfig.doNotSendUselessEntityPackets && isUselessEntityPacket(packet1)) packet1 = null; // Prismarine
                 }
 
                 if ((this.trackDelta || this.entity.hasImpulse || this.entity instanceof LivingEntity && ((LivingEntity) this.entity).isFallFlying()) && this.tickCount > 0) {
@@ -247,6 +248,22 @@ public class ServerEntity {
 
     }
 
+    // Prismarine start
+    private boolean isUselessEntityPacket(Packet<?> possibleUselessEntityPacket) {
+        if (possibleUselessEntityPacket instanceof ClientboundMoveEntityPacket) {
+            ClientboundMoveEntityPacket p = (ClientboundMoveEntityPacket) possibleUselessEntityPacket;
+            if (possibleUselessEntityPacket instanceof ClientboundMoveEntityPacket.Pos) {
+                return p.getXa() == 0 && p.getYa() == 0 && p.getZa() == 0;
+            } else if (possibleUselessEntityPacket instanceof ClientboundMoveEntityPacket.PosRot) {
+                return p.getXa() == 0 && p.getYa() == 0 && p.getZa() == 0 && p.getxRot() == 0 && p.getyRot() == 0;
+            } else if (possibleUselessEntityPacket instanceof ClientboundMoveEntityPacket.Rot) {
+                return p.getxRot() == 0 && p.getyRot() == 0;
+            }
+        }
+        return false;
+    }
+    // Prismarine end
+
     public void removePairing(ServerPlayer player) {
         this.entity.stopSeenByPlayer(player);
         player.connection.send(new ClientboundRemoveEntitiesPacket(new int[]{this.entity.getId()}));
diff --git a/src/main/java/net/prismarineteam/prismarine/PrismarineConfig.java b/src/main/java/net/prismarineteam/prismarine/PrismarineConfig.java
index 58a6a66bf8039a5b3777e18a4c6fdac09d39617e..7c9ac6144d058186b539cc867c91892e4a080ae5 100644
--- a/src/main/java/net/prismarineteam/prismarine/PrismarineConfig.java
+++ b/src/main/java/net/prismarineteam/prismarine/PrismarineConfig.java
@@ -215,4 +215,9 @@ public class PrismarineConfig {
         useJettPackAsyncExecutor = getBoolean("settings.jettpack-patches.use-jettpack-async-executor", useJettPackAsyncExecutor);
         useYatopiaAsyncNBTCache = getBoolean("settings.jettpack-patches.use-yatopia-async-nbt-cache", useYatopiaAsyncNBTCache);
     }
+    
+    public static boolean doNotSendUselessEntityPackets = true;
+    private static void doNotSendUselessEntityPackets() {
+        doNotSendUselessEntityPackets = getBoolean("settings.do-not-send-useless-entity-packets", doNotSendUselessEntityPackets);
+    }
 }
\ No newline at end of file
